synapse/python_dependencies.py

```diff
@@ -171,10 +174,16 @@ def check_requirements(for_feature=None):
             else:
                 errors.append("Needed %s but it was not installed" % (dependency,))
 
-    if not for_feature:
-        # Check the optional dependencies are up to date. We allow them to not be
+    if for_feature is None:
+        # Check if the optional runtime dependencies are up to date. We allow them to not be
         # installed.
-        OPTS = sum(CONDITIONAL_REQUIREMENTS.values(), [])  # type: List[str]
+        OPTS = set().union(
+            *(
+                set(reqs)
+                for key, reqs in CONDITIONAL_REQUIREMENTS.items()
+                if key in ALL_RUNTIME_REQUIREMENT_COMPONENTS or dev
+            )
+        )  # type: Set[str]
```

Comment:
@richvdh this is two compounded changes;

- Change `OPTS` to be a flattened set, to deduplicate any elements, and have potential faster iteration.
- if `dev == False`, only check default (runtime) dependencies, if `dev == True`, also check development dependencies, this is useful as an addition in `__main__` down below, which does `dev=True`, making an invocation of the file (`python -m synapse.python_dependencies`) automatically also check those development dependencies.